src/outputs/rcp.c

```diff
@@ -0,0 +1,1464 @@
+rcp_device_add(const char *device_name, const char *address, int port)
+{
+  struct output_device *device;
+  uint64_t id;
+  int ret;
+
+  /* possible if this was mdns/avahi that there is a timeout/cached entry is no
+   * longer resolvable;  ie avahi caches a roku name, the roku is pwoered down,
+   * cache entry still exists but now no address
+   */ 
+  if (!address)
+    {
+      DPRINTF(E_DBG, L_RCP, "Attempt to add '%s' without address (possible mdns timeout)\n", device_name ? device_name : "<no name>");
+      return;
+    }
+
+  id = djb_hash(device_name, strlen(device_name));
+  if (!id)
+    {
+      DPRINTF(E_LOG, L_RCP, "Could not hash RCP name for id (%s)\n", device_name);
+      return;
+    }
+
+  CHECK_NULL(L_RCP, device = calloc(1, sizeof(struct output_device)));
+
+  device->id = id;
+  device->name = strdup(device_name);
+  device->type = OUTPUT_TYPE_RCP;
+  device->type_name = outputs_name(device->type);
+  device->v4_address = strdup(address);
+  device->v4_port = port;
+  DPRINTF(E_INFO, L_RCP, "Adding RCP output device '%s' at '%s'\n", device_name, address);
+
+  ret = player_device_add(device);
+  if (ret < 0)
+    outputs_device_free(device);
+}
+
+static void
+rcp_device_remove(const char *device_name, const char *address, int port)
+{
+  struct output_device *device;
+  uint64_t id;
+  int ret;
+
+  if (address || port != -1)
+    {
+      DPRINTF(E_DBG, L_RCP, "Attempt to remove '%s' with non null address (%s)/port(%d)\n", device_name, address, port);
+      return;
+    }
+
+  id = djb_hash(device_name, strlen(device_name));
+  if (!id)
+    {
+      DPRINTF(E_LOG, L_RCP, "Could not hash RCP name for id (%s)\n", device_name);
+      return;
+    }
+
+  CHECK_NULL(L_RCP, device = calloc(1, sizeof(struct output_device)));
+
+  device->id = id;
+  device->name = strdup(device_name);
+  device->type = OUTPUT_TYPE_RCP;
+  device->type_name = outputs_name(device->type);
+  device->v4_port = 1;
+  DPRINTF(E_INFO, L_RCP, "Removing RCP output device '%s'\n", device_name);
+
+  ret = player_device_remove(device);
+  if (ret < 0)
+    outputs_device_free(device);
+}
+
+static void
+rcp_mdns_device_cb(const char *name, const char *type, const char *domain, const char *hostname, int family, const char *address, int port, struct keyval *txt)
+{
+  bool exclude;
+
+  /* $ avahi-browse -vrt  _roku-rcp._tcp
+	Server version: avahi 0.7; Host name: foo.local
+	E Ifce Prot Name                       Type                 Domain
+	+  eth0 IPv4 SoundBridge              _roku-rcp._tcp       local
+	=  eth0 IPv4 SoundBridge              _roku-rcp._tcp       local
+	   hostname = [SoundBridge.local]
+	   address = [192.168.0.3]
+	   port = [5555]
+	   txt = []
+	: Cache exhausted
+	: All for now
+   */
+
+  exclude = cfg_getbool(cfg_gettsec(cfg, "rcp", name), "exclude");
+  DPRINTF(E_DBG, L_RCP, "Event for %sRCP/SoundBridge device '%s' (address %s, port %d)\n", exclude ? "excluded " : "", name, address, port);
+  
+  if (exclude)
```

Comment:
Consider just returning after logging. Then you don't need the else and don't need indentitation of the happy path.